### PR TITLE
[DCA] Add gc goroutine to HPAWatcher to handle deleted hpas.

### PR DESCRIPTION
**Pull request by devonboyer**
### What does this PR do?

Added
- `gc` in `HPAWatcherClient` to delete HPAs configured by `hpa_watcher_gc_period`
- Added a lock to functions in `configMapStore` since the `DatadogProvider` can try to read while the `HPAWatcherClient` updates and the `configMapStore` functions always work on a local copy.

### Motivation

The previous delete logic relied on splitting the keys of the data stored in the configmap to figure out what to delete. This was error prone because there is no safe way to split the key and parse the parts since the key components like `MetricName` and `Namespace` can contain the key delimeter.

This also solves the problem of HPAs being deleted while the DCA is not running.

### TODO

- [x] Run e2e test for locally
- [x] Sanity check that deletes happen

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -313,7 +313,7 @@ func init() {
 	Datadog.BindEnv("cluster_agent.cmd_port")
 	Datadog.BindEnv("cluster_agent.kubernetes_service_name")
 	BindEnvAndSetDefault("hpa_watcher_polling_freq", 10)
-
+	BindEnvAndSetDefault("hpa_watcher_gc_period", 30) // 30 seconds
```
Comment (devonboyer) [inline pkg/config/config.go]: Metrics for deleted HPAs will remain in the configmap for this long.

File: pkg/util/kubernetes/hpa/hpa.go [hunk 2]
```diff
@@ -215,6 +250,7 @@ func (c *HPAWatcherClient) processHPAs(added, modified []*autoscalingv2.Horizont
 					HPA: custommetrics.ObjectReference{
 						Name:      hpa.Name,
 						Namespace: hpa.Namespace,
+						UID:       string(hpa.UID),
```
Comment (devonboyer) [inline pkg/util/kubernetes/hpa/hpa.go]: `UID` is used by `HPAWatcherClient.gc`.

File: pkg/util/kubernetes/hpa/hpa.go [hunk 3]
```diff
@@ -159,11 +164,41 @@ func (c *HPAWatcherClient) Start() {
 				// Updating the metrics against Datadog should not affect the HPA pipeline.
 				// If metrics are temporarily unavailable for too long, they will become `Valid=false` and won't be evaluated.
 				c.updateExternalMetrics()
+			case <-gcTicker.C:
+				if !leaderEngine.IsLeader() {
```
Comment (CharlyF) [inline pkg/util/kubernetes/hpa/hpa.go]: How confident are we that the local stores across replicas are consistent ?
If the leader updates the shared store (cm) but one of the replicas is not yet up to date with the previous state, will that work as expected ? Are we in an eventually consistent pattern ?

File: pkg/util/kubernetes/hpa/hpa.go [hunk 4]
```diff
@@ -147,9 +155,6 @@ func (c *HPAWatcherClient) Start() {
 					if err = c.processHPAs(added, modified); err != nil {
 						log.Errorf("Could not update the external metrics in the store: %s", err.Error())
 					}
-					if err = c.removeEntryFromStore(deleted); err != nil {
-						log.Errorf("Could not delete the external metrics in the store: %s", err.Error())
-					}
 				}
 			// Ticker to run the refresh process for the stored external metrics
 			case <-tickerHPARefreshProcess.C:
```
Comment (CharlyF) [inline pkg/util/kubernetes/hpa/hpa.go]: Personal note: This ticker freq has to be lower than the update from the common store.
`hpa_watcher_polling_freq` is at 10s, and `external_metrics_provider.polling_freq` is at 30s so it's good.

**Review by CharlyF — COMMENTED** [review]
I left a few comments - Essentially this is great but I want to make sure we are not making unnecessary calls to the API Server or duplicating some calls.

Also, just so the workflow is clear:
- `run` is done by the leader, it watches HPAs. 
-  HPAs are processed, and added to the global store (CM). This is done every 10 seconds. So 
* 1 Watch + potentially a call to the CM
- `updateExternalMetrics` takes care of writing to the global store
- we run the `gc` every 30 sec (could be bumped?) to update the local store of the leader and the global store

In parralele:
- The HPA Controller calls `ListAllExternalMetrics` every 30 sec, leading to a `Resync` of the local store
- This Resync -> `resync` gets the values from the global store.
* 1 Get CM every 30 sec.
 
Main question is, will there be a consistency issue with the replicas not being consistent but behind the APIService ?

File: pkg/clusteragent/custommetrics/store.go [hunk 5]
```diff
@@ -141,46 +142,45 @@ func (c *configMapStore) SetExternalMetricValues(added []ExternalMetricValue) er
...
 }
 
 // ListAllExternalMetricValues returns the most up-to-date list of external metrics from the configmap.
 // Any replica can safely call this function.
 func (c *configMapStore) ListAllExternalMetricValues() ([]ExternalMetricValue, error) {
-	var metrics []ExternalMetricValue
-	if err := c.getConfigMap(); err != nil {
+	c.mu.Lock()
+	err := c.getConfigMap()
+	c.mu.Unlock()
```
Comment (mfpierre) [inline pkg/clusteragent/custommetrics/store.go]: I'd prefer doing `defer c.mu.Unlock()` above and dropping the `RLock` below, this seems more readable/less error prone and the performance is not critical here

Comment (devonboyer) [reply pkg/clusteragent/custommetrics/store.go]: That's true. If you'd prefer that I can do that. `ListAllExternalMetricValues` is also the function that is called the most times through the DCA so I was trying to pretend that that was a real performance improvement.

Comment (mfpierre) [reply pkg/clusteragent/custommetrics/store.go]: the "possibly concurrently" hints that we should go for readability here 😁

**Review by CharlyF — APPROVED** [review]
🔍 👨‍🚀

**Pull request merged**